CS-1385: SC Synthetic UH: Calculate parameters in user interface

KipSullivan opened the pull request:
Initial commit - there are still a couple of outstanding small issues to be addressed. Wanted to get your feedback on the current progress with #1385 and see if there is room for improvement/suggestions to improve/change it.

- Rearranged 'Select Time Of Concentration' to be toward the top, along with compacting the bottom half of the form (fields like 'Drainage Area' and 'Peak Rate Factor' are combined into a 2x3 grid
     - Still wondering if maybe the 'Time of Concentration, Tc' should be moved higher up the form as well.
![newpagelayout](https://user-images.githubusercontent.com/76274215/183503242-11cea9d1-d019-498b-a137-111bc7b21df0.PNG)

- 'Calculate Missing Parameters' button on the bottom of the page calculates most of the parameter fields automatically, using the endpoints listed in #1385
![autocalculate](https://user-images.githubusercontent.com/76274215/183503411-a303910c-2d05-4f07-b893-c5cf4c62f35d.PNG)

- More descriptive warning below the 'Calculate Missing Parameters' button to help user identify what needs to be filled out
![calculationrequirements](https://user-images.githubusercontent.com/76274215/183503741-0f45989e-5204-4254-ab94-ce8141dfddea.PNG)



Things yet to be done:
- If the drop downs are left blank, there isn't an orange outline indicating that they are blank, which are present for other text/number input fields. 
- Potentially move 'Time of Concentration, Tc' field higher up the form as well.
- More descriptive error handling if one of the parameters returned from the 'calculateMissingParametersSCSUH1' endpoint is missing or an unexpected value

Commit:
initial endpoint connections

Commit:
#1385 added ability to calculate parameters in SC Synthetic UH

Commit:
#1385 fixed merge conflicts

amedenblik reviewed (changes requested):
This is coming together nicely!

- [x] Can you edit the calculated values, so they don't have so many digits? There is an existing function `.toUSGSValue()` that will limit the values to the appropriate number of significant digits. 

![image](https://user-images.githubusercontent.com/40237491/183512420-93fca91c-41e8-4f8b-9198-a27a56854fc2.png)

- [x] Let's move "Select a CN Modification Method" below "Select a Rainfall Distribution Curve", since the CN Modification Method isn't one of the parameters that is calculated automatically.

- [x] I'm not sure the Rainfall Distribution Curve is being calculated correctly. For the point I tried (34.01380519814052, -80.99063158035278), the service returned `rainfall_distribution_curve_letter: "B"` but the drop-down showed "Type II" (should be "NOAA B").

- [x] If any values are manually entered, we don't want to overwrite those when we click "Calculate Missing Parameters". For example, if I enter Length and Slope for the Lag Time Equation, these values should not be overwritten when I click "Calculate Missing Parameters".

- [x] I think the "Channelized Flow - Storm Sewer Segments" are not getting translated correctly to the payload for the services. I entered 4 segments with pipe materials CMP, PVC, Concrete, and Steel, but the payload included pipe materials Aluminum, PVC, Concrete, and Concrete. 

- [x] We could rearrange the inputs like this, to better match how things are organized in the spreadsheet. I am open to other ideas, though! We may get additional feedback from SCDOT too.
![image](https://user-images.githubusercontent.com/40237491/183517712-10985f8b-30f4-43f3-af2e-7769f3ed0771.png)

- [x] "If the drop downs are left blank, there isn't an orange outline indicating that they are blank, which are present for other text/number input fields." -- if you can get this working, that would be cool! It's okay if not. 

- [x] "More descriptive error handling if one of the parameters returned from the 'calculateMissingParametersSCSUH' endpoint is missing or an unexpected value" this seems useful. 

Let us know if you could use any assistance or have further questions.

Commit:
#1385 rearranged fields, fixed rainfall distribution curve/channelized flow storm segment, added outline to empty dropdown

KipSullivan commented:
- Rearranged calculated inputs, while also limiting the number of significant digits of the calculations and not overriding any pre-existing values
![dontoverwrite](https://user-images.githubusercontent.com/76274215/184157505-e1b51661-cdab-4816-83bb-5965b50169f5.PNG)

- Adding yellow outline to empty drop down fields
![warningondropdowns](https://user-images.githubusercontent.com/76274215/184157640-c2a9c318-8392-46e6-8918-6bca5b209970.PNG)

- Fixed "Channelized Flow - Storm Sewer Segments" translation issue for endpoint calculation
- Fixed "Rainfall Distribution Curve" not being calculated correctly
- Moved "Select a CN Modification Method" below "Select a Rainfall Distribution Curve"
- More descriptive error handling for params that aren't calculated correctly in the "calculateMissingParametersSCSUH" endpoint

harper-wavra commented on src/Controllers/SCStormRunoffController.ts:
```diff
@@ -62,6 +63,9 @@ module StreamStats.Controllers {
         public lagTimeLength: number;
         public lagTimeSlope: number;
+        public syntheticParameters;
+        //
```
Cleanup unused comment

harper-wavra commented on src/Controllers/SCStormRunoffController.ts:
```diff
@@ -884,6 +900,179 @@ module StreamStats.Controllers {
+                                "X-warning": true
+                            };
+                            let formmatedSegments = this.getFormattedFlowSegments();
+                            // master data endpoing
```
spelling

harper-wavra commented on src/Controllers/SCStormRunoffController.ts:
```diff
@@ -884,6 +900,179 @@ module StreamStats.Controllers {
+                                if (param.code.toLowerCase() == 'lfplength' && this._selectedTimeOfConcentration?.value == 2 && !this.lagTimeSlope) this.lagTimeSlope =  param.value;
+                            });
+
+                            //** */ Other SCStormModelingServices data
```
comment is formatted a bit weird

harper-wavra reviewed (changes requested):
I think the calculate missing parameters button is working really well, and from my testing I think you were able to fix all the functionality notes that Andrea addressed. I just made a couple code clean-up comments. 

- [x] Andrea and I talked and we think we should move the CN modification Method (selected manually) above the rainfall distribution curve (can be calculated automatically). That way all the values that _have_ to be entered by the user at the top.

amedenblik reviewed (commented):
I don't have any additional comments beyond what Harper included in her review!

Commit:
#1385 moved CN mod method above rainful distr. num and removed unnecessary comments

KipSullivan commented:
Just cleaned up the code a bit following the suggestions from @harper-wavra - let me know if I missed anything!
- Moved 'CN Modification Method' drop down above the 'Rainfall Distribution Curve' drop down
- Removed unnecessary comments in code and fixed spelling on variables

harper-wavra commented:
Those changes all look good, thanks!
